src/main/java/org/springframework/retry/support/RetryTemplateBuilder.java

```diff
@@ -0,0 +1,431 @@
+package org.springframework.retry.support;
+
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.springframework.classify.BinaryExceptionClassifier;
+import org.springframework.classify.BinaryExceptionClassifierBuilder;
+import org.springframework.retry.RetryListener;
+import org.springframework.retry.RetryPolicy;
+import org.springframework.retry.backoff.BackOffPolicy;
+import org.springframework.retry.backoff.ExponentialBackOffPolicy;
+import org.springframework.retry.backoff.ExponentialRandomBackOffPolicy;
+import org.springframework.retry.backoff.FixedBackOffPolicy;
+import org.springframework.retry.backoff.NoBackOffPolicy;
+import org.springframework.retry.backoff.UniformRandomBackOffPolicy;
+import org.springframework.retry.policy.AlwaysRetryPolicy;
+import org.springframework.retry.policy.BinaryExceptionClassifierRetryPolicy;
+import org.springframework.retry.policy.CompositeRetryPolicy;
+import org.springframework.retry.policy.MapRetryContextCache;
+import org.springframework.retry.policy.MaxAttemptsRetryPolicy;
+import org.springframework.retry.policy.RetryContextCache;
+import org.springframework.retry.policy.TimeoutRetryPolicy;
+import org.springframework.util.Assert;
+
+/**
+ * Fluent API to configure new instance of RetryTemplate.
+ * All time units are in milliseconds.
+ * For detailed description of each builder method - see it's doc.
+ *
+ * <p>
+ * Examples:
+ * <pre>{@code
+ * RetryTemplate.newBuilder()
+ *      .maxAttempts(10)
+ *      .exponentialBackoff(100, 2, 10000)
+ *      .retryOn(IOException.class)
+ *      .traversingCauses()
+ *      .build();
+ *
+ * RetryTemplate.newBuilder()
+ *      .fixedBackoff(10)
+ *      .withinMillis(3000)
+ *      .build();
+ *
+ * RetryTemplate.newBuilder()
+ *      .infiniteRetry()
+ *      .retryOn(IOException.class)
+ *      .uniformRandomBackoff(1000, 3000)
+ *      .build();
+ * }</pre>
+ *
+ * <p>
+ * The builder provides the following defaults:
+ * <ul>
+ * <li> max attempts = 3 (initial + 2 retries) Inherited from {@link MaxAttemptsRetryPolicy} </li>
+ * <li> retry only on {@link Exception} and it's subclasses, without traversing causes
+ *      Inherited from {@link BinaryExceptionClassifier#getDefaultClassifier()} </li>
+ * <li> no backoff (retry immediately) </li>
+ * <li> throw an {@link org.springframework.retry.ExhaustedRetryException} on exhausted,
```

Comment:
This it true only for stateful retry, thinking about removing this comment, or removing whole `throwLastExceptionOnExhausted` and `retryContextCache` methods from the builder API (because they are rarely used)

Reply:
Finally, removed retryContextCache and throwLastExceptionOnExhausted from builder.

Reply:
I don't like that much to be honest. Rarely used is a judgement call.

Reply:
Dave, let me check: do you also think that there is no sense to have `retryContextCache` and `throwLastExceptionOnExhausted` customization in builder API?